File: binary-search/minimum-scanner-range-for-warehouse-aisle-labels-1785351659/java/Solution.java
import java.util.*;

/*
Problem Title: Minimum Scanner Range for Warehouse Aisle Labels

Problem Description:
A warehouse has several aisle labels placed along a straight corridor. The positions of the labels are given in a sorted integer array labels, where labels[i] is the position of the i-th label on the corridor. You are also given an integer k, the number of handheld scanners available. Each scanner can be placed at any real-valued position and can read every label whose distance from the scanner is at most R, where R is the scanner's reading range. All scanners use the same range.

Your task is to find the minimum integer value R such that all labels can be covered using at most k scanners.

A scanner covers a continuous interval [x - R, x + R], so once a scanner is placed, it may cover multiple nearby labels. You may choose scanner positions optimally. Return the smallest possible integer R.

This problem should be solved efficiently for large inputs. A common approach is to binary search the answer R and greedily check whether all labels can be covered with at most k scanners.

Constraints:
- 1 <= labels.length <= 100000
- 0 <= labels[i] <= 1000000000
- labels is sorted in non-decreasing order
- 1 <= k <= labels.length
- Return an integer answer

Example 1:
Input: labels = [1, 2, 8, 12, 17], k = 2
Output: 4
Explanation: Place one scanner at position 4 to cover labels in [0, 8], covering 1, 2, and 8. Place the second scanner at position 13 to cover [9, 17], covering 12 and 17. Range 3 is not enough, so the minimum range is 4.

Example 2:
Input: labels = [0, 5, 6, 7, 20], k = 3
Output: 1
Explanation: One scanner at 0 covers label 0. One scanner at 6 covers labels 5, 6, and 7. One scanner at 20 covers label 20. Therefore range 1 is sufficient, and range 0 would not cover both 5 and 7 with a single scanner.
*/

public class Solution {

    /**
     * Finds the minimum integer scanner range R such that all label positions can be covered
     * using at most k scanners.
     *
     * Core idea:
     * 1. Binary search on the answer R.
     * 2. For a fixed R, greedily determine how many scanners are needed.
     * 3. If k scanners are enough, try a smaller R; otherwise try a larger R.
     *
     * Why the greedy check works:
     * - Suppose we start from the leftmost uncovered label at position p.
     * - To maximize coverage, place the scanner as far right as possible while still covering p.
     *   That means placing it at position p + R.
     * - Then this scanner covers every label up to p + 2R.
     * - This is optimal for covering the maximum number of labels starting from p.
     *
     * @param labels sorted array of label positions along the corridor
     * @param k maximum number of scanners available
     * @return the smallest integer range R that allows all labels to be covered
     * Time complexity: O(n log D), where n is labels.length and D is labels[n-1] - labels[0]
     * Space complexity: O(1)
     */
    public int minimumScannerRange(int[] labels, int k) {
        int n = labels.length;

        // If we have at least as many scanners as labels, each label can get its own scanner.
        // In that case, range 0 is enough.
        if (k >= n) {
            return 0;
        }

        // The answer must lie between:
        // - 0 (minimum possible range)
        // - labels[last] - labels[first] (safe upper bound)
        //
        // Why is this upper bound safe?
        // With one scanner, a range equal to the full spread is definitely enough,
        // because we can place the scanner somewhere to cover all labels.
        int left = 0;
        int right = labels[n - 1] - labels[0];

        // Standard binary search for the minimum feasible value.
        while (left < right) {
            int mid = left + (right - left) / 2;

            // If range mid is sufficient, try to shrink the answer.
            if (canCover(labels, k, mid)) {
                right = mid;
            } else {
                // Otherwise, mid is too small, so we must go larger.
                left = mid + 1;
            }
        }

        return left;
    }

    /**
     * Checks whether all labels can be covered using at most k scanners,
     * when every scanner has range r.
     *
     * Greedy strategy:
     * - Start from the first uncovered label.
     * - Let that label be at position start.
     * - Place a scanner at position start + r.
     *   This is the furthest-right placement that still covers start.
     * - That scanner then covers every label up to start + 2*r.
     * - Skip all labels within that covered interval.
     * - Repeat until all labels are covered or we exceed k scanners.
     *
     * This greedy approach is optimal for a fixed r because each scanner is used
     * to cover as many labels as possible from left to right.
     *
     * @param labels sorted array of label positions
     * @param k maximum number of scanners allowed
     * @param r candidate scanner range to test
     * @return true if all labels can be covered with at most k scanners, otherwise false
     * Time complexity: O(n)
     * Space complexity: O(1)
     */
    public boolean canCover(int[] labels, int k, int r) {
        int n = labels.length;
        int scannersUsed = 0;
        int index = 0;

        // Process labels from left to right.
        while (index < n) {
            scannersUsed++;

            // If we already used too many scanners, this range is not sufficient.
            if (scannersUsed > k) {
                return false;
            }

            // The current leftmost uncovered label.
            long start = labels[index];

            // If we place the scanner at start + r, it covers up to start + 2*r.
            // Use long to avoid overflow when positions are large.
            long coveredUntil = start + 2L * r;

            // Skip every label that lies within this scanner's coverage.
            //
            // Detailed reasoning:
            // - labels[index] is definitely covered because it equals start.
            // - Any later label <= coveredUntil is also covered.
            // - The first label > coveredUntil becomes the next uncovered label.
            while (index < n && labels[index] <= coveredUntil) {
                index++;
            }
        }

        // If we finished the loop, all labels were covered using at most k scanners.
        return true;
    }

    /**
     * Demonstrates the solution on the sample test cases from the problem statement.
     *
     * @param args command-line arguments (not used)
     * @return nothing
     * Time complexity: O(1) for the fixed demo calls, excluding the called algorithm
     * Space complexity: O(1)
     */
    public static void main(String[] args) {
        Solution solution = new Solution();

        int[] labels1 = {1, 2, 8, 12, 17};
        int k1 = 2;
        int result1 = solution.minimumScannerRange(labels1, k1);
        System.out.println("Example 1 result: " + result1);
        System.out.println("Expected: 4");

        int[] labels2 = {0, 5, 6, 7, 20};
        int k2 = 3;
        int result2 = solution.minimumScannerRange(labels2, k2);
        System.out.println("Example 2 result: " + result2);
        System.out.println("Expected: 1");

        // Additional quick sanity checks for beginners:
        int[] labels3 = {5};
        int k3 = 1;
        System.out.println("Single label result: " + solution.minimumScannerRange(labels3, k3));
        System.out.println("Expected: 0");

        int[] labels4 = {1, 10};
        int k4 = 1;
        System.out.println("Two labels, one scanner result: " + solution.minimumScannerRange(labels4, k4));
        System.out.println("Expected: 5");
    }
}

File: binary-search/minimum-scanner-range-for-warehouse-aisle-labels-1785351659/python/solution.py
"""
Title: Minimum Scanner Range for Warehouse Aisle Labels

Problem Description:
A warehouse has several aisle labels placed along a straight corridor. The positions
of the labels are given in a sorted integer array `labels`, where `labels[i]` is the
position of the i-th label on the corridor. You are also given an integer `k`, the
number of handheld scanners available. Each scanner can be placed at any real-valued
position and can read every label whose distance from the scanner is at most `R`,
where `R` is the scanner's reading range. All scanners use the same range.

Your task is to find the minimum integer value `R` such that all labels can be covered
using at most `k` scanners.

A scanner covers a continuous interval [x - R, x + R], so once a scanner is placed,
it may cover multiple nearby labels. You may choose scanner positions optimally.
Return the smallest possible integer `R`.

This problem should be solved efficiently for large inputs. A common approach is to
binary search the answer `R` and greedily check whether all labels can be covered
with at most `k` scanners.

Constraints:
- 1 <= labels.length <= 100000
- 0 <= labels[i] <= 1000000000
- labels is sorted in non-decreasing order
- 1 <= k <= labels.length
- Return an integer answer

Example 1:
Input: labels = [1, 2, 8, 12, 17], k = 2
Output: 4

Example 2:
Input: labels = [0, 5, 6, 7, 20], k = 3
Output: 1
"""

from typing import List


class Solution:
    def can_cover(self, labels: List[int], k: int, radius: int) -> bool:
        """
        Check whether all labels can be covered using at most k scanners
        when every scanner has reading range `radius`.

        Args:
            labels: Sorted list of label positions along the corridor.
            k: Maximum number of scanners available.
            radius: Candidate scanner range being tested.

        Returns:
            True if all labels can be covered with at most k scanners, else False.

        Time complexity:
            O(n), where n is the number of labels.

        Space complexity:
            O(1), ignoring input storage.
        """
        # This variable counts how many scanners we have used so far.
        scanners_used: int = 0

        # This index walks through the labels from left to right.
        i: int = 0
        n: int = len(labels)

        # We greedily cover labels from left to right.
        #
        # Why greedy works:
        # - Suppose labels[i] is the leftmost uncovered label.
        # - To cover it while also maximizing how far right we can reach,
        #   the best scanner placement is centered at labels[i] + radius.
        # - That scanner then covers the interval:
        #       [labels[i], labels[i] + 2 * radius]
        #   because its left edge is (labels[i] + radius - radius) = labels[i]
        #   and its right edge is (labels[i] + radius + radius) = labels[i] + 2R.
        # - Any other placement that still covers labels[i] cannot extend farther
        #   to the right than this.
        #
        # Therefore, each time we encounter the next uncovered label, we place one
        # scanner in the best possible way and skip every label it covers.
        while i < n:
            scanners_used += 1

            # If we already used more than k scanners, this radius is not enough.
            if scanners_used > k:
                return False

            # The current scanner starts by covering the leftmost uncovered label.
            # We place it at labels[i] + radius, so the farthest right label it can
            # cover is any label <= labels[i] + 2 * radius.
            cover_right: int = labels[i] + 2 * radius

            # Move forward while labels are still inside this scanner's coverage.
            while i < n and labels[i] <= cover_right:
                i += 1

        # If we finished all labels using at most k scanners, the radius works.
        return True

    def minScannerRange(self, labels: List[int], k: int) -> int:
        """
        Find the minimum integer scanner range needed to cover all labels
        using at most k scanners.

        Args:
            labels: Sorted list of label positions along the corridor.
            k: Number of scanners available.

        Returns:
            The smallest integer radius R such that all labels are covered.

        Time complexity:
            O(n log D), where n is the number of labels and
            D = labels[-1] - labels[0].

        Space complexity:
            O(1), ignoring input storage.
        """
        # Edge case:
        # If we have at least as many scanners as labels, we can place one scanner
        # directly on each label, so radius 0 is enough.
        if k >= len(labels):
            return 0

        # Binary search over the answer.
        #
        # Why binary search is valid:
        # - If a radius R works, then any larger radius also works.
        # - If a radius R does not work, then any smaller radius also does not work.
        # This monotonic property makes binary search appropriate.
        #
        # Lower bound:
        # - 0 is the smallest possible integer radius.
        #
        # Upper bound:
        # - labels[-1] - labels[0] is always enough, because one scanner with that
        #   radius can certainly cover the entire span if placed appropriately.
        left: int = 0
        right: int = labels[-1] - labels[0]

        # Standard binary search for the first feasible radius.
        while left < right:
            mid: int = (left + right) // 2

            # Test whether this candidate radius is sufficient.
            if self.can_cover(labels, k, mid):
                # If it works, try to find an even smaller valid radius.
                right = mid
            else:
                # If it does not work, we must increase the radius.
                left = mid + 1

        # At loop end, left == right and points to the minimum feasible radius.
        return left


if __name__ == "__main__":
    solution = Solution()

    # Example 1
    labels1: List[int] = [1, 2, 8, 12, 17]
    k1: int = 2
    result1: int = solution.minScannerRange(labels1, k1)
    print("Example 1:")
    print(f"labels = {labels1}, k = {k1}")
    print(f"Minimum scanner range = {result1}")
    print("Expected = 4")
    print()

    # Example 2
    labels2: List[int] = [0, 5, 6, 7, 20]
    k2: int = 3
    result2: int = solution.minScannerRange(labels2, k2)
    print("Example 2:")
    print(f"labels = {labels2}, k = {k2}")
    print(f"Minimum scanner range = {result2}")
    print("Expected = 1")
    print()

    # Additional quick sanity checks
    labels3: List[int] = [5]
    k3: int = 1
    result3: int = solution.minScannerRange(labels3, k3)
    print("Additional Test 1:")
    print(f"labels = {labels3}, k = {k3}")
    print(f"Minimum scanner range = {result3}")
    print("Expected = 0")
    print()

    labels4: List[int] = [1, 10, 20, 30]
    k4: int = 2
    result4: int = solution.minScannerRange(labels4, k4)
    print("Additional Test 2:")
    print(f"labels = {labels4}, k = {k4}")
    print(f"Minimum scanner range = {result4}")